Add code region folding to CodeEdit

jmb462 opened the pull request:
Here is an implementation of my code region folding proposal.
Closes https://github.com/godotengine/godot-proposals/issues/3366

# Region code folding

While developping my game, I feel unconfortable navigating in scripts that may be quite long and contains many methods.
I miss a way to organize section of code and navigate throw it more easily.

As discuss in the proposal, I know that when misused, code folding can be consider as antipattern by some devs but I think it will help for those who need it and have a good use of it. User still have the choice to use it or not. Positive reactions under the proposal convince me to open this PR.

# How does it works

Code region are defined by the use of two common region tag (can be used out of the box in VSCode).

Opening tag is ``#region name_of_the_region``
Closing tag is ``#endregion``
To be language agnostic, as CodeEdit can be used for any language, the first symbol depends of the comment delimiters of CodeEdit. So ``#`` will be used in Godot code editor while developping with GDScript language.

# Region creation

Region can be created manualy by writing opening and closing tag around the code that we want to include in this region.

![image](https://user-images.githubusercontent.com/3649998/224577524-fd811ec5-6588-4f9b-853f-1049275f2b11.png)


I've also added a context menu on text selection that do the stuff for us. We just need to modify the default name with this way.

![region](https://user-images.githubusercontent.com/3649998/224577599-3b831459-f5d1-435d-b2f5-d500fc8a3dfb.gif)

Edit : Keyboard shorcut ``Alt + R`` has been added to create region.

 # Region highlighting

When a region is folded, its line is highlighted and a special gutter is shown.
While unfolded, there is no highlight and classic folding gutter is shown.

![image](https://user-images.githubusercontent.com/3649998/224577733-86619a70-f546-4256-a1fa-6f411f933f98.png)

As a colorblind, the default color of line background and gutter icon is maybe not the best choice. 

I choose the existing folder icon for region gutter. Of course a special icon could be designed. I don't really have Inkscape skill. I found that folder icon was not a bad temporay choice.

Clicking on the gutter icon (folder) will unfold the region.

Region color highlighting can be customized in editor settings under ``text_editor/theme/highlighting/folded_code_region_color``.

# Unit tests

I've added some unit tests in CodeEdit to check region detection, creation and folding.

# Methods overview panel (EDIT : THIS PART IS DELAYED AND WILL BE ADDED IN ANOTHER PR)

The ItemList of method overview has been replaced by a Tree.

Nothing seems changed if not using regions :

![image](https://user-images.githubusercontent.com/3649998/224578116-7cc5a9e9-21e5-4b26-8131-3fffe016749b.png)

When a region is defined in the script, it will appeared as a root element in the tree with (again) a folder icon :

![image](https://user-images.githubusercontent.com/3649998/224578067-cbc19f0f-1637-403f-86f2-19d2aa893e2f.png)

I choose to not synchronize the folding of region in the script and in the method overview. Because i think we could want to fold script without folding the corresponding methods in the overview.

Clicking on the region name in the overview send you at the corresponding line in the script and unfold the region.
The region is also unfolded if we click on a method in that region.

Filtering works on methods and region names.

If alphabetical sort is toggled on, the root elements of the overview (region and methods that are not in a region) are sorted at first, then, inside each region, methods are also sorted.

Script order :
![image](https://user-images.githubusercontent.com/3649998/224578411-380f4548-6f1a-4705-a929-a8767ef64308.png)

Alphabetical order : 
![image](https://user-images.githubusercontent.com/3649998/224578434-1d2fe472-9456-4ce3-9c71-259e68c32d92.png)

### Video preview

https://user-images.githubusercontent.com/3649998/224578706-9bcef45c-a4cb-442b-aea9-33048dce660f.mp4

jmb462 committed:
Add code region folding to CodeEdit

seppoday commented:
Can we add shortcut to that feature?

jmb462 commented:
> Is there a possibility to add shortcut?

I didn't think about it at first, but yes, as the other options in context menu, it should be possible to add a keyboard shortcut for that. I will look at it.

seppoday commented:
> > Is there a possibility to add shortcut?
> 
> For creating a region ? I didn't think about it at first, but yes, as the other options in context menu, it should be possible to add a keyboard shortcut for that. I will look at it.
> 
> (It should also be possible to add shortcut to quickly go to the next / previous region.)

I mean, dunno if it will be usefull for anyone. We have shortcut for creating comments, so I figured that it might be usefull to add shortcut also here. Some people like to basically only use keyboard.

xchrix commented:
It's great for someone like me who has to write long scripts.

jmb462 commented:
Updated to use CodeEdit theme cache as it was conflicting with master since YuriSizov PR merge.

Paulb23 commented on scene/gui/code_edit.cpp:
```diff
@@ -1536,31 +1558,45 @@ void CodeEdit::fold_line(int p_line) {
+			}
+		}
+		set_line_background_color(p_line, theme_cache.folded_code_region_color);
+		queue_redraw();
```
No need to call `queue_redraw` here.

Paulb23 commented on scene/gui/code_edit.cpp:
```diff
@@ -1650,6 +1717,29 @@ TypedArray<int> CodeEdit::get_folded_lines() const {
 
+/* Code region */
+String CodeEdit::get_code_region_opening_tag() const {
+	return _get_code_region_tag("region");
```
Should we make the tag a property?

Paulb23 commented on scene/gui/code_edit.cpp:
```diff
@@ -1626,6 +1666,33 @@ void CodeEdit::unfold_all_lines() {
 	_unhide_all_lines();
 }
 
+void CodeEdit::create_code_region() {
```
This looks like it needs support adding for multiple carets

jmb462 replied:
Not sure to understand what you mean here.

If multiselection is active, right mouse click unselect all the selections and so we do not have access to the "Create region" menu item.

So maybe we just need to early return if multiselection is active so that the keyboard shortcut has no effect too.

Or did you mean an other thing ?

Paulb23 commented on scene/gui/code_edit.cpp:
```diff
@@ -1488,6 +1494,22 @@ bool CodeEdit::can_fold_line(int p_line) const {
+				WARN_PRINT_ONCE(vformat("Nested code regions are not supported. (Lines %s and %s)", p_line, next_line));
+				return false;
+			}
+		}
```
Can we early return here?

jmb462 replied:
I think so because if we allow to fold a nested region start line, it will fold until the next endregion even there is another region start between these two lines, it's sure that we do not want that.

Alternative solution is to allow nested region and keep track of a count of tags for detecting corresponding region tag.

Paulb23 commented on scene/gui/code_edit.h:
```diff
@@ -185,6 +185,9 @@ class CodeEdit : public TextEdit {
 	int code_hint_xpos = -0xFFFF;
 
+	/* Code region */
+	String _get_code_region_tag(String p_tag) const;
```
Should be grouped under folding.

Could it also be made a `const String &p_tag`?

Paulb23 commented on scene/gui/code_edit.h:
```diff
@@ -377,12 +382,19 @@ class CodeEdit : public TextEdit {
 	void fold_line(int p_line);
 	void unfold_line(int p_line);
 	void fold_all_lines();
+	void create_code_region();
```
Similar here, needs grouping with the rest of the methods.

Paulb23 commented on scene/gui/code_edit.h:
```diff
@@ -377,12 +382,19 @@ class CodeEdit : public TextEdit {
 	bool is_line_folded(int p_line) const;
 	TypedArray<int> get_folded_lines() const;
 
+	/* Code region */
```
Using `open / close` sounds better to me then `opening / closing`.

jmb462 replied:
> Reviewed `/scene` changes.
> 
> Not sure on making the editor side quite so advanced yet as I would like to refactor that first, might be worth splitting those changes out into a separate PRs.

Thanks for you review, I'm working on it with your comments. For the editor part, maybe we could remove from this PR the replacement of the item list by the tree (if that's the part of editor you want to refactor first) and keep only "Create code region" menu item and shortcut.

For delimiters, it will now choose the first one-line comment delimiter (not pushed to github yet). I think it's good enough. All languages seems to have a one line delimiter. Supporting an endkey comment delimiter would make code more complex for a use case that seems more than rare. If someone with such a language try to create a region, it will just warn that there we can't without a oneline comment defined.

Paulb23 commented on scene/gui/code_edit.cpp:
```diff
@@ -2818,6 +2908,15 @@ TypedArray<String> CodeEdit::_get_delimiters(DelimiterType p_type) const {
 
+/* Code Region */
+String CodeEdit::_get_code_region_tag(String p_tag) const {
+	TypedArray<String> comment_delimiters = get_comment_delimiters();
```
Might be an idea to check if it's line `line_only`, or search for one that is, else it will need a closing symbol.

Paulb23 reviewed (changes requested):
Reviewed `/scene` changes.

Not sure on making the editor side quite so advanced yet as I would like to refactor that first, might be worth splitting those changes out into a separate PRs.

jmb462 commented:
@Paulb23 I've updated the PR thanks to your review.

1°) Methods overview has been revert to ItemList instead of Tree. It will not show anymore the code region. I will propose a separate PR if this one is merged.

2°) Code region tag (start / end) are now property. They are only updated when comment delimiters are modified (added, removed or cleared)

3°) Code region methods have been grouped with code folding methods in the header files.

4°) opening / closing vocabulary has been replaced with start / end to match comment delimiters vocabulary.

5°) Each time comments delimiters are modified, the region tags are updated to select the first available delimiters that is line_only without end delimiter. For now, it doesn't support langage with no one line comment. As I said in an other comment, it's seems good enough because it seems that it's a common have of all languages.

6°) Disable create code region via shortcut if several carrets / selections are active to avoid unwanted behaviour. In master, right click on mouse while having multiple selection will unselected all, so we can already not create region by menu when multiple selections are actives.

7°) Updated unit tests with checks for good update of region tags when modifying comment delimiters.

Paulb23 commented on tests/scene/test_code_edit.h:
```diff
@@ -2691,6 +2691,132 @@ TEST_CASE("[SceneTree][CodeEdit] folding") {
+	SceneTree::get_singleton()->get_root()->add_child(code_edit);
+	code_edit->grab_focus();
+
+	SUBCASE("[CodeEdit] region folding") {
```
Some addition tests:

- create_code_region
-- No selection
-- No valid comment delimiter
-- mutli-carets

Case when we have an existing code region then the tag changes.

Paulb23 reviewed (changes requested):
> 1°) Methods overview has been reverted to ItemList instead of Tree. It will not show anymore the code region. I will propose a separate PR if this one is merged.

Thanks :)

> 6°) Disable create code region via shortcut if several carrets / selections are active to avoid unwanted behaviour. In master, right click on mouse while having multiple selection will unselected all, so we can already not create region by menu when multiple selections are actives.

This shouldn't be the case, right clicking with multiple carets doesn't remove any selections for me, perhaps a bug?
![image](https://user-images.githubusercontent.com/6584330/233847692-7753e7a6-6a8d-44f1-8d6f-783bb27f29f1.png)

Do we know how other editors handle this operation with multiple carets? If we do limit it would be worth mentioning it in the documentation. As we should go ahead and bind these methods to the API as well.

Few other minor changes otherwise looks good.

jmb462 commented:
> As we should go ahead and bind these methods to the API as well.

@Paulb23 I'm finilizing the PR. Not sure of what you meen here ? Do you mean expose code region function of CodeEdit to GDscript ?

jmb462 commented:
@Paulb23 PR Updated

1°) Code region creation now support multiselections
2°) Added more unit tests
3°) Methods exposed to GDScript

Paulb23 commented on scene/gui/code_edit.cpp:
```diff
@@ -1756,6 +1806,74 @@ TypedArray<int> CodeEdit::get_folded_lines() const {
+	_update_code_region_tags();
+}
+
+bool CodeEdit::is_line_code_region_start(int p_line) const {
```
Should check `ERR_FAIL_INDEX_V(p_line, get_line_count(), false);` same in `is_line_code_region_end`

Paulb23 commented on scene/gui/code_edit.cpp:
```diff
@@ -1756,6 +1806,74 @@ TypedArray<int> CodeEdit::get_folded_lines() const {
+}
+
+void CodeEdit::set_code_region_tags(const String &p_start, const String &p_end) {
+	ERR_FAIL_COND_MSG(p_start == p_end, "Starting and ending region tags cannot be identical.");
```
Do we need to disallow empty as well? Or is that intended to disable this behaviour? If the latter would add a note to the documentation.

Paulb23 commented on scene/gui/code_edit.cpp:
```diff
@@ -2872,6 +3022,9 @@ void CodeEdit::_remove_delimiter(const String &p_start_key, DelimiterType p_type
 		}
 		break;
 	}
+	if (p_type == DelimiterType::TYPE_COMMENT) {
+		_update_code_region_tags();
+	}
```
This should be placed inside the loop before break.

Paulb23 commented on scene/gui/code_edit.h:
```diff
@@ -116,6 +116,11 @@ class CodeEdit : public TextEdit {
 
 	/* Line Folding */
 	bool line_folding_enabled = false;
+	String code_region_start = "";
+	String code_region_end = "";
```
Sounds a bit odd, `code_region_[start|end]_string` `code_region_[start|end]_text`?

Paulb23 commented on scene/gui/code_edit.cpp:
```diff
@@ -1756,6 +1806,74 @@ TypedArray<int> CodeEdit::get_folded_lines() const {
+	begin_complex_operation();
+	// Adding start and end region tags
+	int first_region_start = -1;
+	for (int caret_idx : get_caret_index_edit_order()) {
```
There's a edge case here if there are overlapping selections like this:

![image](https://github.com/godotengine/godot/assets/6584330/78828aee-2294-46f5-803a-fdd055d623dc)

In the other multi-caret code, we'd combine them into one region, we should probably do something similar here.

jmb462 replied:
```// Merge selections if next selection starts on the same line the previous one ends
	Vector<int> caret_edit_order = get_caret_index_edit_order();
	for (int i = 1; i < caret_edit_order.size(); i++) {
		int current_caret = caret_edit_order[i - 1];
		int next_caret = caret_edit_order[i];
		if (get_selection_from_line(current_caret) == get_selection_to_line(next_caret)) {
			select(get_selection_from_line(next_caret), get_selection_from_column(next_caret), get_selection_to_line(current_caret), get_selection_to_column(current_caret), next_caret);
			remove_caret(current_caret);
		}
	}
```

Trying to implents that. This code does the job, but not sure it's the good way to do that. Is there a better way ?

Paulb23 replied:
Can do something like below, or take a look at `void CodeTextEditor::move_lines_up()`:
```
	Vector<int> caret_edit_order = get_caret_index_edit_order();
	Vector<int> carets_to_remove;
	for (int i = 0; i < caret_edit_order.size(); i++) {
		int current_caret = caret_edit_order[i];
		
		int to_line = get_selection_to_line(i, c);
		int from_line = get_selection_from_line(i, c);
	
		// Find next non-overlapping caret.		
		for (int j = i; j < caret_edit_order.size(); j++) {
			if (!has_selection(j)) {
				if (get_caret_column(j) != from_line) {
					break;
				}
				carets_to_remove.append(j);
				i++;
				continue;
			}
			
			if (get_selection_to_line(j) == from_line) {
				from_line = get_selection_to_line(j);
				carets_to_remove.append(j);
				i++;
				continue;
			}
			break;
		}
		
		// handle region
	}
	
	// Sort and remove backwards to preserve indexes.
	carets_to_remove.sort();
	for (int i = carets_to_remove.size() - 1; i >= 0; i--) {
		remove_caret(carets_to_remove[i]);
	}
```

jmb462 replied:
My implementation works in all case I can test. It seems to be sufficient here and more simple to understand.
Do you see any problem with it ?

Paulb23 reviewed (commented):
Tested locally seems good, few nit inline hopefully the last pass.

sepTN reviewed (approved):
I have been wanting region folding in GDScript. Thanks for writing this, been using it on my godot fork. It does seem to do the job.

sepTN commented:
This is the default colorings 
![image](https://github.com/godotengine/godot/assets/13846022/b2b4160c-f7bc-418a-a097-39179a5a3554)

But I'd like to have my region just like any other comments so I set the alpha to zero. But now the folding icons gone since it also has the same alpha of zero
![image](https://github.com/godotengine/godot/assets/13846022/ed8a0fde-5f64-4678-9e56-2f840dabd99f)

Perhaps it will be better if you add two separate setting of this in editor

Calinou commented:
> Perhaps it will be better if you add two separate setting of this in editor

Instead of adding another setting, we can probably `max()` the icon's alpha with a value such as `0.5`. This ensures it's always visible.

jmb462 commented:
> > Perhaps it will be better if you add two separate setting of this in editor
> 
> Instead of adding another setting, we can probably `max()` the icon's alpha with a value such as `0.5`. This ensures it's always visible.

Done

Paulb23 commented on scene/gui/code_edit.cpp:
```diff
@@ -1763,6 +1815,89 @@ TypedArray<int> CodeEdit::get_folded_lines() const {
+		int next_caret = caret_edit_order[i];
+		if (get_selection_from_line(current_caret) == get_selection_to_line(next_caret)) {
+			select(get_selection_from_line(next_caret), get_selection_from_column(next_caret), get_selection_to_line(current_caret), get_selection_to_column(current_caret), next_caret);
+			remove_caret(current_caret);
```
Method looks good. `remove_caret` will change the remaining `caret` indexes, should update `caret_edit_order` or remove them in a separate loop.

jmb462 replied:
Carets are now deleted in a separate loop.

AThousandShips commented on scene/gui/code_edit.cpp:
```diff
@@ -1601,6 +1609,27 @@ bool CodeEdit::can_fold_line(int p_line) const {
 		return false;
 	}
 
+	/* Check for code region. */
```
```suggestion
	// Check for code region.
```
Use single line comments for these cases

Paulb23 commented on scene/gui/code_edit.cpp:
```diff
@@ -1763,6 +1815,93 @@ TypedArray<int> CodeEdit::get_folded_lines() const {
+			caret_to_remove.append(current_caret);
+		}
+	}
+	for (const int &c : caret_to_remove) {
+		remove_caret(c);
+	}
```
Need to do this:
```
// Sort and remove backwards to preserve indexes.
carets_to_remove.sort();
for (int i = carets_to_remove.size() - 1; i >= 0; i--) {
	remove_caret(carets_to_remove[i]);
}
```

jmb462 replied:
Thanks. Done.

Paulb23 reviewed (approved):
Nice work, Code LGTM!

Can add a fancier icon later if we wish

Calinou reviewed (dismissed):
Tested locally, it works as expected.

I noticed two UX issues though:

- Contrast ratio for the region text is very low when collapsed, due to the region highlight being so bright. It also makes the highlight quite distracting:

![image](https://github.com/godotengine/godot/assets/180032/aee39dac-45e6-4b82-bd24-4bba83ab7492)

The highlight looks great on light theme though:

![image](https://github.com/godotengine/godot/assets/180032/005e4aae-7b0f-4b4f-a0a8-fb1bb0e83ce7)

- When you have nested regions, the nested region's folding state is not remembered if you unfold the parent region after folding it.

https://github.com/godotengine/godot/assets/180032/a2d8fe61-5826-4199-98d8-3a9a0effd225

I suggest tweaking the dark theme region highlight color to be less opaque, then this should be good to merge. The second issue is less critical and can be addressed in a future PR.

jmb462 commented:
> I suggest tweaking the dark theme region highlight color to be less opaque, then this should be good to merge. The second issue is less critical and can be addressed in a future PR.

@Calinou As a colorblind, I have much trouble to choose a correct color.
Could you please (or someone else) give a default color code that could be fine ? (can be tweak in settings ``text_editor/theme/highlighting/folded_code_region_color``)

Calinou commented:
> @jmb462 Safest option would be to pick one of the already used accent/highlight colors that the syntax highlighter uses.

There's also the question of whether we should keep using a blue hue, as it looks a lot like the text selection color. Maybe we should go for a purple hue instead?

Calinou commented:
I noticed another oddity in the folding region detection logic. In these 3 situations, I expect folding the second region to be possible in all cases, but it's only feasible in 1 of the cases:

![Screenshot_20230809_114954](https://github.com/godotengine/godot/assets/180032/b151baf3-251f-474e-a064-c208c13ce3cb)

![Screenshot_20230809_115000](https://github.com/godotengine/godot/assets/180032/92eda68f-30b8-40dd-993d-c5cbdee040f4)

![Screenshot_20230809_115008](https://github.com/godotengine/godot/assets/180032/2fdd79e9-bd33-40df-8d50-e18a3e86528b)

Calinou commented:
Regarding the color, I've found this color that works well on both dark and light themes: `Color(0.68, 0.46, 0.77, 0.2)`

Dark theme | Light theme
-|-
![Screenshot_20230809_120146](https://github.com/godotengine/godot/assets/180032/a635a587-7447-4b43-93cb-fb9bfeaa5a6d) | ![image](https://github.com/godotengine/godot/assets/180032/248f3ba7-25bf-4640-a390-2d7777b6a64d)

MewPurPur commented:
Nevermind, I misunderstood something about how the code editor works. I edited the icons in my last message to account for that.

jmb462 commented:
> **Edit:** It works after the editor is restarted.
@Calinou 
I cannot reproduce this issue. The three case are detecting instantly for me.

Calinou commented:
> Do I also have to add these icons to the scene\resources\default_theme folder and so what difference in the svg version of editor/icons. (Names and opacity are differents ??)

If code region folding is available for use in exported projects, then yes, the icons should also be in `scene/resources/default_theme`.

AThousandShips commented on scene/gui/code_edit.cpp:
```diff
@@ -1602,6 +1617,27 @@ bool CodeEdit::can_fold_line(int p_line) const {
+	}
+	if (is_line_code_region_start(p_line)) {
+		int region_level = 0;
+		/* Check if there is a valid end region tag */
```
```suggestion
		// Check if there is a valid end region tag.
```
[Comment style](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide) nitpick

AThousandShips commented on scene/gui/code_edit.cpp:
```diff
@@ -1650,31 +1686,51 @@ void CodeEdit::fold_line(int p_line) {
-		if (end_line == p_line) {
-			for (int i = p_line + 1; i <= line_count; i++) {
-				if ((in_string != -1 && is_in_string(i) == -1) || (in_comment != -1 && is_in_comment(i) == -1)) {
+	// Fold code region
```
```suggestion
	// Fold code region.
```

AThousandShips commented on scene/gui/code_edit.cpp:
```diff
@@ -1650,31 +1686,51 @@ void CodeEdit::fold_line(int p_line) {
+	if (!is_line_code_region_start(p_line)) {
+		if (in_string != -1 || in_comment != -1) {
+			end_line = get_delimiter_end_position(p_line, get_line(p_line).size() - 1).y;
+			/* End line is the same therefore we have a block of single line delimiters. */
```
```suggestion
			// End line is the same therefore we have a block of single line delimiters.
```

AThousandShips commented on scene/gui/code_edit.cpp:
```diff
@@ -1764,6 +1823,95 @@ TypedArray<int> CodeEdit::get_folded_lines() const {
 
+/* Code region */
+void CodeEdit::create_code_region() {
+	// Abort if there is no selected text
```
```suggestion
	// Abort if there is no selected text.
```

AThousandShips commented on scene/gui/code_edit.cpp:
```diff
@@ -1764,6 +1823,95 @@ TypedArray<int> CodeEdit::get_folded_lines() const {
+	if (!has_selection()) {
+		return;
+	}
+	// Check that region tag find a comment delimiter and is valid
```
```suggestion
	// Check that region tag find a comment delimiter and is valid.
```

AThousandShips commented on scene/gui/code_edit.cpp:
```diff
@@ -1764,6 +1823,95 @@ TypedArray<int> CodeEdit::get_folded_lines() const {
+		return;
+	}
+	begin_complex_operation();
+	// Merge selections if selection starts on the same line the previous one ends
```
```suggestion
	// Merge selections if selection starts on the same line the previous one ends.
```

AThousandShips replied:
You have missed this and several others, please pay attention to the hidden review comments

jmb462 replied:
Sorry. I've check all // comments and add missing points.
Some were forgiven by mistake but for some others were on purpose because I thought that if it was not a sentence with a conjugated verb, it doesn't need final period. Now periods added everywhere.

AThousandShips commented on scene/gui/code_edit.cpp:
```diff
@@ -1764,6 +1823,95 @@ TypedArray<int> CodeEdit::get_folded_lines() const {
+		remove_caret(carets_to_remove[i]);
+	}
+
+	// Adding start and end region tags
```
```suggestion
	// Adding start and end region tags.
```

AThousandShips commented on scene/gui/code_edit.cpp:
```diff
@@ -1764,6 +1823,95 @@ TypedArray<int> CodeEdit::get_folded_lines() const {
+		fold_line(from_line);
+	}
+
+	// Select name of the first region to allow quick edit
```
```suggestion
	// Select name of the first region to allow quick edit.
```

AThousandShips commented on tests/scene/test_code_edit.h:
```diff
@@ -2839,6 +2839,194 @@ TEST_CASE("[SceneTree][CodeEdit] folding") {
+	SUBCASE("[CodeEdit] region folding") {
+		code_edit->set_line_folding_enabled(true);
+
+		// Region tag detection
```
```suggestion
		// Region tag detection.
```

AThousandShips commented on tests/scene/test_code_edit.h:
```diff
@@ -2839,6 +2839,194 @@ TEST_CASE("[SceneTree][CodeEdit] folding") {
+		CHECK_FALSE(code_edit->is_line_code_region_end(1));
+		CHECK(code_edit->is_line_code_region_end(2));
+
+		// Region tag customization
```
```suggestion
		// Region tag customization.
```

AThousandShips commented on tests/scene/test_code_edit.h:
```diff
@@ -2839,6 +2839,194 @@ TEST_CASE("[SceneTree][CodeEdit] folding") {
+		CHECK(code_edit->get_code_region_start_tag() == "region");
+		CHECK(code_edit->get_code_region_end_tag() == "endregion");
+
+		// Region creation with selection adds start / close region lines
```
```suggestion
		// Region creation with selection adds start / close region lines.
```

AThousandShips commented on tests/scene/test_code_edit.h:
```diff
@@ -2839,6 +2839,194 @@ TEST_CASE("[SceneTree][CodeEdit] folding") {
+		CHECK(code_edit->get_line(2).contains("line2"));
+		CHECK(code_edit->is_line_code_region_end(3));
+
+		// Region creation without any selection has no effect
```
```suggestion
		// Region creation without any selection has no effect.
```

AThousandShips commented on tests/scene/test_code_edit.h:
```diff
@@ -2839,6 +2839,194 @@ TEST_CASE("[SceneTree][CodeEdit] folding") {
+		code_edit->create_code_region();
+		CHECK(code_edit->get_text() == "line1\nline2\nline3");
+
+		// Region creation with multiple selections
```
```suggestion
		// Region creation with multiple selections.
```

AThousandShips commented on tests/scene/test_code_edit.h:
```diff
@@ -2839,6 +2839,194 @@ TEST_CASE("[SceneTree][CodeEdit] folding") {
+		code_edit->create_code_region();
+		CHECK(code_edit->get_text() == "#region New Code Region\nline1\n#endregion\nline2\n#region New Code Region\nline3\n#endregion");
+
+		// Two selections on the same line create only one region
```
```suggestion
		// Two selections on the same line create only one region.
```

AThousandShips commented on tests/scene/test_code_edit.h:
```diff
@@ -2839,6 +2839,194 @@ TEST_CASE("[SceneTree][CodeEdit] folding") {
+		code_edit->create_code_region();
+		CHECK(code_edit->get_text() == "#region New Code Region\ntest line1\ntest line2\ntest line3\n#endregion");
+
+		// Region tag with // comment delimiter
```
```suggestion
		// Region tag with // comment delimiter.
```

AThousandShips commented on tests/scene/test_code_edit.h:
```diff
@@ -2839,6 +2839,194 @@ TEST_CASE("[SceneTree][CodeEdit] folding") {
+		CHECK(code_edit->is_line_code_region_start(0));
+		CHECK(code_edit->is_line_code_region_end(2));
+
+		// Creating region with no valid one line comment delimiter has no effect
```
```suggestion
		// Creating region with no valid one line comment delimiter has no effect.
```

AThousandShips commented on tests/scene/test_code_edit.h:
```diff
@@ -2839,6 +2839,194 @@ TEST_CASE("[SceneTree][CodeEdit] folding") {
+		code_edit->create_code_region();
+		CHECK(code_edit->get_text() == "line1\nline2\nline3");
+
+		// Choose one line comment delimiter
```
```suggestion
		// Choose one line comment delimiter.
```

AThousandShips commented on tests/scene/test_code_edit.h:
```diff
@@ -2839,6 +2839,194 @@ TEST_CASE("[SceneTree][CodeEdit] folding") {
+		CHECK(code_edit->is_line_code_region_start(0));
+		CHECK(code_edit->is_line_code_region_end(2));
+
+		// Update code region delimiter when removing comment delimiter
```
```suggestion
		// Update code region delimiter when removing comment delimiter.
```

AThousandShips commented on tests/scene/test_code_edit.h:
```diff
@@ -2839,6 +2839,194 @@ TEST_CASE("[SceneTree][CodeEdit] folding") {
+		CHECK(code_edit->is_line_code_region_start(3));
+		CHECK(code_edit->is_line_code_region_end(5));
+
+		// Update code region delimiter when clearing comment delimiters
```
```suggestion
		// Update code region delimiter when clearing comment delimiters.
```

AThousandShips commented on tests/scene/test_code_edit.h:
```diff
@@ -2839,6 +2839,194 @@ TEST_CASE("[SceneTree][CodeEdit] folding") {
+		CHECK_FALSE(code_edit->is_line_code_region_start(0));
+		CHECK_FALSE(code_edit->is_line_code_region_end(2));
+
+		// Fold region
```
```suggestion
		// Fold region.
```

AThousandShips commented on tests/scene/test_code_edit.h:
```diff
@@ -2839,6 +2839,194 @@ TEST_CASE("[SceneTree][CodeEdit] folding") {
+		CHECK(code_edit->is_line_folded(0));
+		CHECK(code_edit->get_next_visible_line_offset_from(1, 1) == 4);
+
+		// Not ending region can't be folded
```
```suggestion
		// Not ending region can't be folded.
```

AThousandShips commented on tests/scene/test_code_edit.h:
```diff
@@ -2839,6 +2839,194 @@ TEST_CASE("[SceneTree][CodeEdit] folding") {
+		CHECK_FALSE(code_edit->can_fold_line(0));
+		ERR_PRINT_ON;
+
+		// Bad nested region can't be folded
```
```suggestion
		// Bad nested region can't be folded.
```

AThousandShips commented on tests/scene/test_code_edit.h:
```diff
@@ -2839,6 +2839,194 @@ TEST_CASE("[SceneTree][CodeEdit] folding") {
+		CHECK(code_edit->can_fold_line(1));
+		ERR_PRINT_ON;
+
+		// Nested region folding
```
```suggestion
		// Nested region folding.
```

AThousandShips commented on tests/scene/test_code_edit.h:
```diff
@@ -2839,6 +2839,194 @@ TEST_CASE("[SceneTree][CodeEdit] folding") {
+		CHECK(code_edit->get_next_visible_line_offset_from(1, 1) == 4);
+		ERR_PRINT_ON;
+
+		// Unfolding a line inside a region unfold whole region
```
```suggestion
		// Unfolding a line inside a region unfold whole region.
```

AThousandShips commented on scene/gui/code_edit.cpp:
```diff
@@ -1764,6 +1823,95 @@ TypedArray<int> CodeEdit::get_folded_lines() const {
+			carets_to_remove.append(current_caret);
+		}
+	}
+	// Sort and remove backwards to preserve indexes.
```
```suggestion
	// Sort and remove backwards to preserve indices.
```

jmb462 commented:
Update done.

@AThousandShips I've modified the comments.

I've used the same icons for default theme as editor icons. It's already very neutral and unicolor. If @MewPurPur want/can modify the editor version icon to better fit default theme requierements but I don't feel the current icon is chocking in exported projects.

akien-mga commented:
Needs a rebase (and then an OK from Calinou ;)).

Calinou commented:
With the latest revision of the PR, the folding icon is very hard to see:

![image](https://github.com/godotengine/godot/assets/180032/627ec80d-3a89-482d-877d-ef80133e2ff7)

Double-check whether the icon SVG is actually opaque. If we modulate it with a transparent color, it should probably be opaque.

MewPurPur commented:
My updated SVGs are no longer translucent, just purr #fff

jmb462 commented:
Stupidly swaped a MAX and MIN call on alpha. I will fixe it today.

jmb462 commented:
@Calinou Checking with a color picker, it's the same color as the line highlight. Seems to be an optical illusion.

Should we choose another color (or boost icon opacity to counteract this feeling ?

![image](https://github.com/godotengine/godot/assets/3649998/12a671c7-127f-46c0-90ac-e946d76bb5b8)
![image](https://github.com/godotengine/godot/assets/3649998/3530c710-445f-4c34-a867-95a70de97f8a)

Calinou commented:
> Should we choose another color (or boost icon opacity to counteract this feeling ?

Yes, I think we can increase the icon opacity.

jmb462 commented:
> Yes, I think we can increase the icon opacity.

Is it ok with a minimum alpha of 0.4 ? 

![image](https://github.com/godotengine/godot/assets/3649998/66ec3f07-9442-4bda-8f3c-0d96e3341fbb)
![image](https://github.com/godotengine/godot/assets/3649998/363c0191-2e2f-4561-827d-225153e1ecb7)

Calinou commented:
> Is it ok with a minimum alpha of 0.4 ?

Looks good :slightly_smiling_face:

AThousandShips commented on tests/scene/test_code_edit.h:
```diff
@@ -2839,6 +2839,194 @@ TEST_CASE("[SceneTree][CodeEdit] folding") {
+		CHECK(code_edit->is_line_code_region_end(5));
+		code_edit->set_code_region_tags("region", "endregion");
+
+		// Setting identical start and end region tags should failed.
```
```suggestion
		// Setting identical start and end region tags should fail.
```

AThousandShips commented on tests/scene/test_code_edit.h:
```diff
@@ -2839,6 +2839,194 @@ TEST_CASE("[SceneTree][CodeEdit] folding") {
+		CHECK(code_edit->is_line_folded(0));
+		CHECK(code_edit->get_next_visible_line_offset_from(1, 1) == 4);
+
+		// Not ending region can't be folded.
```
```suggestion
		// Non ending region can't be folded.
```
Or
```suggestion
		// Region with no end can't be folded.
```

YuriSizov commented:
@Calinou Was the icon the only thing missing, or is there anything else you'd like to have addressed?

AThousandShips commented on doc/classes/CodeEdit.xml:
```diff
@@ -137,6 +137,15 @@
+				Creates a new code region with the selection. At least one single line comment delimiter have to be defined (see [method add_comment_delimiter]).
+				A code region is a part of code that is highlighted when folded and can help organize your script.
+				Code region start and end tags can be customized (see [method set_code_region_tags]).
+				Code region are delimited using start and end tags (respectively [code]region[/code] and [code]endregion[/code] by default) preceded by one line comment delimiter. (eg. [code]#region[/code] and [code]#endregion[/code])
```
```suggestion
				Code regions are delimited using start and end tags (respectively [code]region[/code] and [code]endregion[/code] by default) preceded by one line comment delimiter. (eg. [code]#region[/code] and [code]#endregion[/code])
```

Calinou commented:
> @Calinou Was the icon the only thing missing, or is there anything else you'd like to have addressed?

Everything in the PR looks good on my end now.

jmb462 commented:
New git conflit fixed (EditorStringNames singleton)

AThousandShips commented on scene/gui/code_edit.h:
```diff
@@ -125,6 +125,11 @@ class CodeEdit : public TextEdit {
 
 	/* Line Folding */
 	bool line_folding_enabled = false;
+	String code_region_start_string = "";
+	String code_region_end_string = "";
```
```suggestion
	String code_region_start_string;
	String code_region_end_string;
```
These already default to this.

AThousandShips replied:
I mean not as such, but it's unnecessary, and we usually suggest it elsewhere

jmb462 replied:
Nearly all Strings are explicitly set to "" in this file, that's why I chose to keep the same style.
But ok, I'm doing the modification for my new Strings.

akien-mga commented:
Needs another rebase, and handling the minor style comment from @AThousandShips. Then this should be ready to merge :)

akien-mga commented:
Thanks!

markofjohnson commented:
Great, thank you! Looking forward to the methods overview panel part down the road.
